libparsec/crates/client/src/client/recovery_device.rs

```diff
@@ -0,0 +1,373 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use libparsec_client_connection::{AuthenticatedCmds, ConnectionError, ProxyConfig};
+use libparsec_platform_device_loader::{
+    get_default_key_file, save_device, PlatformExportRecoveryDeviceError,
+    PlatformImportRecoveryDeviceError, SaveDeviceError,
+};
+use libparsec_platform_storage::certificates::{GetCertificateError, PerTopicLastTimestamps};
+use libparsec_protocol::authenticated_cmds::v4::device_create;
+use libparsec_types::{
+    AvailableDevice, DeviceLabel, DeviceSaveStrategy, LocalDevice, SecretKeyPassphrase,
+};
+
+use crate::{
+    greater_timestamp, CertifPollServerError, CertificateBasedActionOutcome,
+    GreaterTimestampOffset, InvalidCertificateError,
+};
+use libparsec_types::prelude::*;
+
+use super::Client;
+
+pub async fn export_recovery_device(
+    client: &Client,
+    device_label: DeviceLabel,
+) -> Result<(SecretKeyPassphrase, Vec<u8>), ExportRecoveryDeviceError> {
+    let (passphrase, data, recovery_device) =
+        libparsec_platform_device_loader::export_recovery_device(&client.device, device_label)
+            .await?;
+
+    // save recovery device
+    let outcome =
+        register_new_device(client.cmds.clone(), recovery_device, client.device.clone()).await?;
+
+    let latest_known_timestamps = match outcome {
+        CertificateBasedActionOutcome::LocalIdempotent => return Ok((passphrase, data)), // not sure how it could happen though
```

Comment:
Your comment is right: `CertificateBasedActionOutcome::LocalIdempotent` is never occurs here !

(this variant is used when the action you want to do has already been done, so if you download the latest certificates you end up in the same state than if your action had succeeded. This works typically when revoking a user or giving a new role to a user in a workspace)

Here things are different: if somebody else create a device with the same `DeviceID`[^1], then you cannot carry on since it means the private keys for this device you own are invalid.

[^1]: Of course `DeviceID` is a UUID so in practice this should never occur under normal circumstances !

I think it's better to define your own outcome type (that could be private) instead of relying on `CertificateBasedActionOutcome`, this way you don't end up with this additional variant that is of no use here.